File: programmers/kakao/level4/동굴_탐험.py
def solution(n, path, order):
    graph = [[] for _ in range(n)]
    for a, b in path:
        graph[a].append(b)
        graph[b].append(a)
        
    pre = [0]*n
    for a, b in order:
        pre[b] = a
    post = [0]*n
    
    visited = [False]*n

    def dfs(node):
        stack = [node]
        visited[node] = True
        while stack:
            cur = stack.pop()
            for nxt in graph[cur]:
                if post[cur]:
                    dfs(post[cur])
                if not visited[pre[cur]]:
                    post[pre[cur]] = cur
                    continue
                if not visited[nxt]:
                    stack.append(nxt)
                    visited[nxt] = True
    dfs(0)

    return sum(visited) == n


n = 9
path = [[0,1],[0,3],[0,7],[8,1],[3,6],[1,2],[4,7],[7,5]]
order = [[8,5],[6,7],[4,1]]
print(solution(n, path, order)) # True

n = 9
path =[[8,1],[0,1],[1,2],[0,7],[4,7],[0,3],[7,5],[3,6]]
order = [[4,1],[5,2]]
print(solution(n, path, order)) # True

n = 9
path = [[0,1],[0,3],[0,7],[8,1],[3,6],[1,2],[4,7],[7,5]]
order = [[4,1],[8,7],[6,5]]
print(solution(n, path, order)) # False

File: Java/programmers/kakao/level4/동굴탐험.java
package CodeTest.Java.programmers.kakao.level4;

import java.util.*;

public class 동굴탐험 {
    private static List<List<Integer>> graph;
    private static int[] p_visited;
    private static int[] n_visited;
    private static boolean[] visited;

    public boolean solution(int n, int[][] path, int[][] order) {
        graph = new ArrayList<>();
        for (int i=0; i<n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] p : path) {
            graph.get(p[0]).add(p[1]);
            graph.get(p[1]).add(p[0]);
        }

        p_visited = new int[n];
        
        for (int[] o : order) {
            p_visited[o[1]] = o[0];
        }

        n_visited = new int[n];
        visited = new boolean[n];

        dfs(0);

        int cnt = 0;
        for (int i=0; i<n; i++) {
            if (visited[i]) cnt++;
        }

        return cnt == n;
    }

    private static void dfs(int node) {
        Stack<Integer> stk = new Stack<>();
        stk.push(node);
        visited[node] = true;

        while (!stk.isEmpty()) {
            int cur = stk.pop();

            if (n_visited[cur] != 0) dfs(n_visited[cur]);
            
            if (!visited[p_visited[cur]]) {
                n_visited[p_visited[cur]] = cur;
                continue;
            }
            
            for (Integer nxt : graph.get(cur)) {    
                if (!visited[nxt]) {
                    stk.add(nxt);
                    visited[nxt] = true;
                }
            }
        }
    }
}
